File: frontend/src/shared/api/documents.ts
import { apiClient } from "./client";

export type ExportFormat = "csv" | "xlsx" | "pdf";
export type DocType = "incoming" | "outgoing" | "internal";

export interface DocTypeConfig {
  doc_type: DocType;
  prefix: string;
  suffix: string;
  include_department: boolean;
  seq_padding: number;
  reset_period: "never" | "yearly" | "monthly";
}

export interface DocTypeConfigUpdate {
  prefix?: string;
  suffix?: string;
  include_department?: boolean;
  seq_padding?: number;
  reset_period?: "never" | "yearly" | "monthly";
}

export interface RegisterDocumentRequest {
  entity_id: string;
  record_id: string;
  doc_type: DocType;
  department_code?: string | null;
  filing_case_id?: string | null;
}

export interface DocumentRegistration {
  id: string;
  entity_id: string;
  record_id: string;
  doc_type: DocType;
  department_code: string | null;
  registration_no: string;
  filing_case_id: string | null;
  registered_by: string | null;
  registered_at: string;
}

export interface FilingCase {
  id: string;
  parent_id: string | null;
  index_code: string;
  title: string;
  retention_years: number | null;
  storage_location: string | null;
  status: "open" | "closed" | "archived";
  close_by: string | null;
  closed_at: string | null;
  responsible_user_id: string | null;
  created_at: string;
}

export interface FilingCaseCreate {
  index_code: string;
  title: string;
  parent_id?: string | null;
  retention_years?: number | null;
  storage_location?: string | null;
  close_by?: string | null;
  responsible_user_id?: string | null;
}

export interface FilingCaseUpdate {
  title?: string;
  retention_years?: number | null;
  storage_location?: string | null;
  status?: "open" | "closed" | "archived";
  close_by?: string | null;
  responsible_user_id?: string | null;
}

export async function listDocTypeConfigs(appId: string): Promise<DocTypeConfig[]> {
  const { data } = await apiClient.get<DocTypeConfig[]>(`/apps/${appId}/documents/doc-types`);
  return data;
}

export async function updateDocTypeConfig(appId: string, docType: DocType, body: DocTypeConfigUpdate): Promise<DocTypeConfig> {
  const { data } = await apiClient.put<DocTypeConfig>(`/apps/${appId}/documents/doc-types/${docType}`, body);
  return data;
}

export async function registerDocument(appId: string, body: RegisterDocumentRequest): Promise<DocumentRegistration> {
  const { data } = await apiClient.post<DocumentRegistration>(`/apps/${appId}/documents/register`, body);
  return data;
}

export async function listRegistrations(appId: string, params?: { doc_type?: DocType }): Promise<DocumentRegistration[]> {
  const { data } = await apiClient.get<DocumentRegistration[]>(`/apps/${appId}/documents/registrations`, { params });
  return data;
}

export async function listFilingCases(appId: string): Promise<FilingCase[]> {
  const { data } = await apiClient.get<FilingCase[]>(`/apps/${appId}/documents/filing-cases`);
  return data;
}

export async function createFilingCase(appId: string, body: FilingCaseCreate): Promise<FilingCase> {
  const { data } = await apiClient.post<FilingCase>(`/apps/${appId}/documents/filing-cases`, body);
  return data;
}

export async function updateFilingCase(appId: string, caseId: string, body: FilingCaseUpdate): Promise<FilingCase> {
  const { data } = await apiClient.patch<FilingCase>(`/apps/${appId}/documents/filing-cases/${caseId}`, body);
  return data;
}

export async function deleteFilingCase(appId: string, caseId: string): Promise<void> {
  await apiClient.delete(`/apps/${appId}/documents/filing-cases/${caseId}`);
}

export async function exportFilingCases(appId: string, format: ExportFormat = "csv"): Promise<void> {
  const response = await apiClient.get(
    `/apps/${appId}/documents/filing-cases/export`,
    { params: { format }, responseType: "blob" },
  );

  const mimeTypes: Record<ExportFormat, string> = {
    csv: "text/csv",
    xlsx: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    pdf: "application/pdf",
  };

  const blob = new Blob([response.data as BlobPart], { type: mimeTypes[format] });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `filing_cases.${format}`;
  document.body.appendChild(a);
  a.click();
  a.remove();
  URL.revokeObjectURL(url);
}

File: backend/app/schemas/documents.py
"""Document registrar + номенклатура дел schemas (ТЗ 3.10)."""
import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

DOC_TYPES = ("incoming", "outgoing", "internal")
RESET_PERIODS = ("never", "yearly", "monthly")
CASE_STATUSES = ("open", "closed", "archived")


class DocTypeConfigRead(BaseModel):
    doc_type: str
    prefix: str
    suffix: str
    include_department: bool
    seq_padding: int
    reset_period: str
    model_config = {"from_attributes": True}


class DocTypeConfigUpdate(BaseModel):
    prefix: str | None = Field(default=None, max_length=32)
    suffix: str | None = Field(default=None, max_length=32)
    include_department: bool | None = None
    seq_padding: int | None = Field(default=None, ge=1, le=10)
    reset_period: str | None = Field(default=None, pattern=r"^(never|yearly|monthly)$")


class RegisterDocumentRequest(BaseModel):
    entity_id: uuid.UUID
    record_id: uuid.UUID
    doc_type: str = Field(pattern=r"^(incoming|outgoing|internal)$")
    department_code: str | None = Field(default=None, max_length=32)
    filing_case_id: uuid.UUID | None = None


class DocumentRegistrationRead(BaseModel):
    id: uuid.UUID
    entity_id: uuid.UUID
    record_id: uuid.UUID
    doc_type: str
    department_code: str | None
    registration_no: str
    filing_case_id: uuid.UUID | None
    registered_by: uuid.UUID | None
    registered_at: datetime
    model_config = {"from_attributes": True}


class FilingCaseCreate(BaseModel):
    index_code: str = Field(min_length=1, max_length=64)
    title: str = Field(min_length=1, max_length=512)
    parent_id: uuid.UUID | None = None
    retention_years: int | None = Field(default=None, ge=1, le=100)
    storage_location: str | None = Field(default=None, max_length=256)
    close_by: date | None = None
    responsible_user_id: uuid.UUID | None = None


class FilingCaseUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=512)
    retention_years: int | None = Field(default=None, ge=1, le=100)
    storage_location: str | None = Field(default=None, max_length=256)
    status: str | None = Field(default=None, pattern=r"^(open|closed|archived)$")
    close_by: date | None = None
    responsible_user_id: uuid.UUID | None = None


class FilingCaseRead(BaseModel):
    id: uuid.UUID
    parent_id: uuid.UUID | None
    index_code: str
    title: str
    retention_years: int | None
    storage_location: str | None
    status: str
    close_by: date | None
    closed_at: datetime | None
    responsible_user_id: uuid.UUID | None
    created_at: datetime
    model_config = {"from_attributes": True}
